client/landing/stepper/hooks/use-login-url-for-flow.ts

```diff
@@ -0,0 +1,30 @@
+import { addQueryArgs } from '@wordpress/url';
+import { useLocation } from 'react-router';
+import { useLoginUrl } from '../utils/path';
+import { useFlowLocale } from './use-flow-locale';
+import { useSiteData } from './use-site-data';
+import type { Flow } from 'calypso/landing/stepper/declarative-flow/internals/types';
+
+type UseLoginUrlForFlowProps = {
+	flow: Flow;
+};
+
+export function useLoginUrlForFlow( { flow }: UseLoginUrlForFlowProps ): string {
+	const locale = useFlowLocale();
+	const location = useLocation();
+	const { siteId, siteSlug } = useSiteData();
+
+	const redirectTo = addQueryArgs( location.pathname, {
+		...( locale && locale !== 'en' ? { locale } : {} ),
+		...( siteId ? { siteId } : {} ),
+		...( siteSlug ? { siteSlug } : {} ),
+		...Object.fromEntries( new URLSearchParams( location.search ).entries() ),
+	} );
```

Comment:
100% a nit, but I would spread this into 4 `ifs` 
```suggestion
       let redirectTo = location.pathname;
       
       if ( locale && locale !== 'en' ) {
       	const redirectTo = addQueryArgs( location.href, { locale } );
       }
       if ( siteId ) {
       	const redirectTo = addQueryArgs( location.href, { siteId } );
       }
       if ( siteSlug ) {
       	const redirectTo = addQueryArgs( location.href, { siteSlug } );
       }
       if ( siteSlug ) {
       	const redirectTo = addQueryArgs( location.href, { siteSlug } );
       }    
```

And I think if we use `location.href` (vs `pathname`) we don't need to consider the pre-existing params manually.